TYP: Correct type annotation for to_dict.

jsspencer opened the pull request:
The `into` argument of DataFrame.to_dict and Series.to_dict can be either a class or instance of a class of dict; this is covariant - subclasses of dict can also be used. The argument was annotated as `type[dict]` though, so type checkers marked passing initialized objects (required for collections.defaultdict) as an incorrect argument type.

Fix by annotating `into` to take either a subclass of dict or an initialized instance of a subclass of dict.

jsspencer committed:
Correct type annotation for to_dict.

The `into` argument of DataFrame.to_dict and Series.to_dict can be
either a class or instance of a class of dict; this is covariant -
subclasses of dict can also be used. The argument was annotated as
`type[dict]` though, so type checkers marked passing initialized objects
(required for collections.defaultdict) as an incorrect argument type.

Fix by annotating `into` to take either a subclass of dict or an
initialized instance of a subclass of dict.

twoertwein commented on pandas/core/frame.py:
```diff
@@ -1925,7 +1925,7 @@ def _create_data_for_split_and_tight_to_dict(
         self,
         orient: Literal["dict", "list", "series", "split", "tight", "index"] = ...,
-        into: type[dict] = ...,
+        into: type[dict] | dict = ...,
```
Is it possible to use `type[MutableMapping] | MutableMapping`? The documentation says any subclass of `abc.Mapping` (but it needs to be mutable).

jsspencer replied:
Yes, sure. How about (would _typing.py be the best place?)

```
MutableMappingT = TypeVar('MutableMappingT', bound=MutableMapping)
```

and then use `... into: type[MutableMappingT] | MutableMappingT) -> MutableMappingT:`

This should then specify the return type matches the type given for `into` (rather than any `MutableMapping` subclass)

twoertwein replied:
Yes, that sounds good!

jsspencer replied:
Done, though a bit more involved than I planned -- a generic type with a default triggered https://github.com/python/mypy/issues/3737. The workaround is to use overloads and not type the implementation, but manual tests (with pytype) show that the return type is still correctly inferred:

```
reveal_type(df.to_dict(into={})) # gives dict
reveal_type(df.to_dict(into=collections.defaultdict(list)))  # gives collections.defaultdict[nothing, list]
```

jsspencer committed:
Use generic MutableMapping type for to_dict method.

Unfortunately a generic type annotation with a default triggers an
existing mypy limitation (https://github.com/python/mypy/issues/3737).
The current workaround is to use overloads and then not annotate the
implementation containing the default parameter; this still enables mypy
to deduce correct return types.

Two overloads are added for Series.to_dict, even though they could be
combined using a Union type, as at least two overloads are required for
a single method.

Dr-Irv reviewed (commented):
I'm going to defer to @twoertwein  on reviewing this one

jsspencer committed:
Fix formatting

twoertwein commented on pandas/core/frame.py:
```diff
@@ -1947,9 +1948,9 @@ def to_dict(
+        into=dict,
         index: bool = True,
-    ) -> dict | list[dict]:
+    ):
```
Can you add the annotations for the non-overload definition? If it triggers a mypy error within the function, it is okay to add an ignore comment.

twoertwein replied:
Should now work for all cases:

```py
import pandas as pd
from typing import MutableMapping

class MyDict(dict): ...

def test(into: MutableMapping):
    # MutableMapping
    reveal_type(pd.DataFrame([]).to_dict("dict", into=into))
    reveal_type(pd.Series([]).to_dict(into=into))

    # dict
    reveal_type(pd.DataFrame([]).to_dict("dict"))
    reveal_type(pd.Series([]).to_dict())

    # MyDict
    reveal_type(pd.DataFrame([]).to_dict("dict", into=MyDict))
    reveal_type(pd.Series([]).to_dict(into=MyDict))
```

twoertwein committed:
return annotation for non-overload

twoertwein committed:
Merge remote-tracking branch 'upstream/main' into to_dict_into_type_annotate

twoertwein committed:
no keyword should return dict

twoertwein committed:
swap overload order to work for dict subclasses that are passed as keywords

Dr-Irv reviewed (approved):
I created https://github.com/pandas-dev/pandas-stubs/issues/784 so we make a similar change in the stubs

twoertwein commented on pandas/core/series.py:
```diff
@@ -1922,21 +1924,40 @@ def keys(self) -> Index:
+    # error: Incompatible default for argument "into" (default has type "type[
+    # dict[Any, Any]]", argument has type "type[MutableMappingT] | MutableMappingT")
+    @deprecate_nonkeyword_arguments(
+        version="3.0", allowed_args=["self"], name="to_dict"
```
Is this okay @mroeschke? It would make it consistent with `DataFrame.to_dict` (and allow the `dict` default case for typing) but it might be a bit "odd" to require keyword-only for a function that takes exactly one argument

mroeschke replied:
Yup the consistency argument makes sense to go forward with this deprecation

twoertwein replied:
Green now

twoertwein committed:
fix tests

mroeschke commented:
Thanks @jsspencer
